### DISCUSSION TITLE
одевания оружия после Transform

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#500] [Q&A]
...

**Comment by Kor-Angar** [comment]
Вот и я так подумал. запоминать можно. но вот как одеть нужное оружие? :-)

я делал одевание нужного оружия для порта готики 1. но там код был слишком сложный для такой простой задачи. https://www.youtube.com/watch?v=lj6A6dCPoFM

может быть на юнион можно как то сделать экипировку конкретного оружия 

***

вот тут положил https://github.com/dosinabox/g2nr_unofficial_update/discussions/478

**Comment by Kor-Angar** [comment]
> ГГ сохраняет экипированное оружие при выходе из состояния превращения (только с Union, спасибо Slavemaster).

**Reply by Kor-Angar** [reply]
@dosinabox  это? 
[ZBETTERTRANSFORM.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10713798/ZBETTERTRANSFORM.zip)
[ZDONTEQUIPBESTWEAPON.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10713799/ZDONTEQUIPBESTWEAPON.zip)

**Reply by dosinabox** [reply]
Новый патчик zDontEquipBestWeapon позволил сделать так:
```cpp
func void B_StopMagicTransform()
{
	if(PlayerIsTransformed == TRUE)
	{
		if(UnionActivated == TRUE)
		{
			Npc_StopAni(hero,"S_RUN");
			AI_PlayAni(hero,"T_TRFSHOOT_2_STAND");
		}
		else
		{
			AI_UnequipWeapons(hero);
		};
```

**Reply by Kor-Angar** [reply]
а это [ZBETTERTRANSFORM.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10713798/ZBETTERTRANSFORM.zip) про что то другое видимо?

**Reply by dosinabox** [reply]
это про правильное снятие эффектов, смену свитков, переключение агрессии

**Reply by Kor-Angar** [reply]
в `Transform_Spell_Logic_Common` можно добавить на `Npc_GetActiveSpellIsScroll`. тогда можно в код добавить руны для тех кто хочет вставить их через марвин.

```D
func int Transform_Spell_Logic_Common(var int manaInvested,var int cost,var int creature)
{
	if (Npc_GetActiveSpellIsScroll(self))
	{
        	cost = SPL_Cost_Scroll;
	};

	if(self.attribute[ATR_MANA] >= cost)
	{
```

**Reply by Kor-Angar** [reply]
а у магии трансформации происходит после каста вызов функции spell_cast_*** ?

у телепортов есть.. может и у трансформации есть какая то общая? :-)
```D
/
/ ----- neu 1.21 Verteiler fьr die Cast-Funcs -------
func void Spell_Cast_Teleport()
{
```

**Reply by dosinabox** [reply]
вроде нет, но можно добавить и для теста выводить что-нибудь на экран

**Reply by Kor-Angar** [reply]
да вот пытался. но что  пока не получается :-)
```D
func void Spell_Cast_TrfDragonSnapper()
{ 
    Hlp_PrintConsole("Spell_Cast_TrfDragonSnapper");
    B_StartMagicTransform(); 
};


func void Spell_Cast_Transform()
{ 
    Hlp_PrintConsole("Spell_Cast_Transform");
    B_StartMagicTransform(); 
};

func void Spell_Cast_TRF()
{ 
    Hlp_PrintConsole("Spell_Cast_TRF");
    B_StartMagicTransform(); 
};
```

**Reply by Kor-Angar** [reply]
млин. я мод неактивировал :-)

**Reply by Kor-Angar** [reply]
мод активировал. а в отладчик всё равно ничего не приходит

**Reply by Kor-Angar** [reply]
таки получилось. Пространство  имён мешало.

```D
func void Spell_Cast_Transform()
{ 
    Hlp_PrintConsole("Spell_Cast_Transform");
    B_StartMagicTransform(); 
};
```

**Reply by Kor-Angar** [reply]
@dosinabox подскажи пожалуйста какую задачу решает введение отдельных инстанций монстров для превращений ?

**Reply by Kor-Angar** [reply]
[ZBETTERTRANSFORM.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10713798/ZBETTERTRANSFORM.zip) что то у меня не работает на Рокфелере. или оно только на инстанциии PC_hero должно работать?

**Comment by Kor-Angar** [comment]
брать экипированное оружие надо в Spell_Logic_Trf****

**Reply by Kor-Angar** [reply]
туплю. в Spell_Cast_Transform тоже можно напрямую по инстанции героя.

```D
    func void Spell_Cast_Transform()
    {   
        PlayerIsTransformed = 1;

        var c_npc her;      her = Hlp_GetNpc (pc_hero);
        var c_item itm;     ItMw_HeroEquipped = Npc_GetEquippedMeleeWeapon(her);

        Hlp_PrintConsole(Str_Format("   Spell_Cast_Transform %s ItMw_Equipped = %s", her.name, ItMw_HeroEquipped.name));
    };
```

**Reply by Kor-Angar** [reply]
инстанции гипотетически можно барть из файла ини

**Comment by Kor-Angar** [comment]
@dosinabox что то у меня герой после загрузки погиб. наверное надо подстраховаться:
```D
func void B_StopMagicTransform()
{
        ...
        if (Hlp_GetInstanceID(self) != Hlp_GetInstanceID(hero))
        {
            self.attribute[ATR_HITPOINTS] = 0;
        };
        ...
};
```

**Reply by dosinabox** [reply]
после загрузки?

**Reply by Kor-Angar** [reply]
да. сохранился, сразу после превращения и потом загрузил игру.

**Reply by dosinabox** [reply]
сохранился во время анимации?

**Reply by dosinabox** [reply]
а героем был PC_Hero?

**Reply by Kor-Angar** [reply]
> а героем был PC_Hero?

da

> сохранился во время анимации?

не помню. у меня было штук 30 загрузок. такой баг был один раз. я сразу добавил условие и продолжил тесты. а уже потом вспомнил и тебе отписался.

**Reply by dosinabox** [reply]
```cpp
self.attribute[ATR_HITPOINTS] = 0;
```
Эту строку можно вообще убрать.

**Comment by Kor-Angar** [comment]
а это герой сразу после превращения :-)


```D
func void :B_RefreshAtInsert()
{
    //if (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(hero))
    if (Npc_IsPlayer(self) == true)
    && (self.guild > GIL_SEPERATOR_HUM)
    {
        Hlp_PrintConsole(Str_Format("B_RefreshAtInsert(zKA_G2_Transform_Weapon) --> %s guild = %i", self.name, self.guild));
        //B_HeroMagicTransform();
        B_StartMagicTransform();

    };
    B_RefreshAtInsert_Old();        
}; 
```

**Comment by Kor-Angar** [comment]
я тут собрал всё касательно трансформации в отдельный мини-мод. соответственно возник момент с хп за уровень.
я правильно опнимаю, что достататочно просты выявить разницу уровней и начислить хп?

```D
func void B_StartMagicTransform()
{
    if (PlayerIsTransformed == 0)
    {  
        PlayerOldLevel = self.level;
```

```D
func void B_StopMagicTransform()
{
    if (PlayerIsTransformed == 1)
    {
     ...
     
        if (hero.level > PlayerOldLevel)
        {
            PlayerOldLevel = hero.level - PlayerOldLevel;
            Hlp_PrintConsole(Str_Format("   B_StopMagicTransform: DeltaLevel %i", PlayerOldLevel));
            
            hero.attribute[ATR_HITPOINTS_MAX] 	+= HP_PER_LEVEL * PlayerOldLevel;
            hero.attribute[ATR_HITPOINTS] 		+= HP_PER_LEVEL * PlayerOldLevel; 

            PlayerOldLevel = hero.level;
        };
        /*
        if (Hlp_GetInstanceID(self) != Hlp_GetInstanceID(hero))
        {
            self.attribute[ATR_HITPOINTS] = 0;
        };
        */

        PlayerIsTransformed = 0;
    };

**Reply by Kor-Angar** [reply]
мне наверное надо еще проверку сделать на то что `PlayerOldLevel > 0` на всякий случай?

**Reply by dosinabox** [reply]
всё норм

**Comment by Kor-Angar** [comment]
@dosinabox я что то так и не понял какой эффект у заклинания разблокировался на 32 уровне?

**Reply by dosinabox** [reply]
ты про что?

**Reply by Kor-Angar** [reply]
ну вот он доходит до 32 уровня в спелллогик и там появляется какой то визуальный эффект. я всё понять не могу какой он :-)

**Reply by dosinabox** [reply]
![Screenshot_4](https://user-images.githubusercontent.com/39790055/221869495-ce864e56-884e-483d-ba78-3b7f551d3297.jpg)

**Reply by Kor-Angar** [reply]
а разве это свечение должно идти не тогда, когда ГГ сидит на корточках и двигает руками влево , вправо?

**Reply by dosinabox** [reply]
не знаю, посмотри как в чистой 1.30

**Reply by Kor-Angar** [reply]
вот эта анимация которую ты добавила когда ГГ превращается в человека, используется для отмены трансформации. в текущей спелллогик оно не появится так как "механика" инвестирования маны как это было в Г1 была заменена на проверку маны на скриптах и у героя не бывает ситуации когда у него не хватит маны на каст.  

Ровно как и  перестать "инвестировать" ману так же не успеешь так time_per_mana = 0. то есть это анимация задействована, а  что бы её увидеть достаточно просто увеличить time_per_mana, что позволит игроку "отменять" инвестирования маны.

```D
// ------ Instanz fьr alle Transform-Sprьche ------
INSTANCE Spell_Transform (C_Spell_Proto)
{
...
	time_per_mana			= 100;
...
};
```

https://www.youtube.com/watch?v=3XU5wYgJYtY
